### DISCUSSION TITLE
Public user sign up

### DISCUSSION THREAD

**Discussion by alsherif-khalaf** [directus/directus#11884] [Archive]
How can I do like that : 
https://strapi.io/blog/a-beginners-guide-to-authentication-and-authorization-in-strapi

It is simple by axios :
```


import axios from 'axios';
axios
  .post('http://localhost:1337/auth/local/register', {
    username: 'Kapman',
    email: 'test@test.com',
    password: 'Password',
  })
  .then(response => {
    console.log('User profile', response.data.user);
    console.log('User token', response.data.jwt);
  })
  .catch(error => {
    console.log('An error occurred:', error.response);
  });


```

I couldn't find any explanation for that in docs .

**Comment by rijkvanzanten** [comment]
Keep in mind that we're not Strapi, so the fact that they have something in their docs doesn't necessarily mean it will work the same in Directus 😉 

That being said, Directus has a create-user endpoint: https://docs.directus.io/reference/system/users/#create-a-user. The above example from Strapi would roughly translate to the following in Directus:

```js
import axios from 'axios';

axios
  .post('http://localhost:8055/users', {
    email: 'test@test.com',
    password: 'Password',
    role: 'b9cad119-c9cc-454c-abf2-36330fb0d360'
  })
  .then(response => {
    console.log('User profile', response.data.data);
  })
  .catch(error => {
    console.log('An error occurred:', error.response);
  });
```

**Reply by alsherif-khalaf** [reply]
Thank you rijk
I donwloaded the example of nextjs with directus ,  I created API BOT  as example and granted him the ability to create users from Directus permissions ,


I used  [API routes ](https://nextjs.org/docs/api-routes/introduction)in nextjs that handle server side functions , it takes request and sends response  :

```js
import { getDirectusClient } from '../../../lib/directus'
// import getConfig from 'next/config'
// const { publicRuntimeConfig, serverRuntimeConfig } = getConfig()
// const { token } = serverRuntimeConfig

async function handler(req, res) {
  const { email, password, firstName, lastName } = req.body

  if (
    !email ||
    !firstName ||
    !lastName ||
    !email.includes('@') ||
    !password ||
    password.trim().length < 7
  ) {
    res.status(422).json({ message: 'You have to fill data correctly' })
    return
  }

  const directus = await getDirectusClient()
  const users = directus.items('directus_users')
  const checkExist = await users.readByQuery({ search: email })

  if (checkExist.data.length) {
    res.status(400).json({ message: 'This e-mail is used', checkExist })
    return
  }

  const result = await users.createOne({
    first_name: firstName,
    last_name: lastName,
    email: email,
    password: password,
    role: '80634947-7f0f-497e-a748-fa9ca93dab2a',
  })

  res.status(201).json({ message: 'success' })

  return result
}

export default handler



```
and I called it from the front side with a simple function  :

```js

    const createUser = async (e, email, password, firstName, lastName) => {
        e.preventDefault()
        console.log(email, password, firstName, lastName)
        const response = await fetch('/api/auth/signup', {
            method: 'POST',
            body: JSON.stringify({ email, password, firstName, lastName }),
            headers: {
                "Content-Type": 'application/json'
            }
        })

        const data = await response.json()

        setErrorMessage(data.message)

        console.log(errorMessage)

        return data;

    }



```
As this way it worked very well ,

but the login process doesn't work well because of Refresh Tokens , I cant understand it 
I created a front-end function that handle login like that : 
```js


    const loginUser = async (e, email, password, firstName, lastName) => {
        e.preventDefault()
        const result = directus.auth.login({ email, password })
            .then(
                () => {
                    setSuccessMessage("Logged In Successfully")
                    redirect()
                }
            )
            .catch((error) => {
                console.error(error);
                setErrorMessage("Incorrect Data")
            });
        return result
    }

```

I can't understand how to implement refresh token after that 

I created a global state in the context to support the "authenticated" value to the all pages of nextjs  : 

```js


if (directus.auth.token) {
  authenticated = true
}

setTimeout(function () {
  const token = directus.auth
    .refresh()
    .then(() => {
      authenticated = true
    })
    .catch((error) => {
      console.log(error)
      authenticated = false
    })
}, 600000)
```

but it gives me error  when trying to logout  using : directus.auth.logout() 

Cookie “refresh_token” has been rejected because it is already expired.

Object { message: "\"refresh_token\" is required in either the JSON payload or Cookie", extensions: {…} }

**Reply by rijkvanzanten** [reply]
I'm not too familiar with NextJS myself, so i'm not sure if there's any gotchas in this approach. @azrikahar do you have any insights in the above?

**Reply by alsherif-khalaf** [reply]
> I'm not too familiar with NextJS myself, so i'm not sure if there's any gotchas in this approach. @azrikahar do you have any insights in the above?

I found little fix here :
https://github.com/directus/directus/issues/4097

```
REFRESH_TOKEN_COOKIE_SAME_SITE: 'none'
REFRESH_TOKEN_COOKIE_SECURE: 'true'
```

This fixed the error when logout
I will test the Refresh Token after that

**Reply by alsherif-khalaf** [reply]
> Thank you rijk I donwloaded the example of nextjs with directus , I created API BOT as example and granted him the ability to create users from Directus permissions ,
> 
> I used [API routes ](https://nextjs.org/docs/api-routes/introduction)in nextjs that handle server side functions , it takes request and sends response :
> 
> ```js
> import { getDirectusClient } from '../../../lib/directus'
> // import getConfig from 'next/config'
> // const { publicRuntimeConfig, serverRuntimeConfig } = getConfig()
> // const { token } = serverRuntimeConfig
> 
> async function handler(req, res) {
>   const { email, password, firstName, lastName } = req.body
> 
>   if (
>     !email ||
>     !firstName ||
>     !lastName ||
>     !email.includes('@') ||
>     !password ||
>     password.trim().length < 7
>   ) {
>     res.status(422).json({ message: 'You have to fill data correctly' })
>     return
>   }
> 
>   const directus = await getDirectusClient()
>   const users = directus.items('directus_users')
>   const checkExist = await users.readByQuery({ search: email })
> 
>   if (checkExist.data.length) {
>     res.status(400).json({ message: 'This e-mail is used', checkExist })
>     return
>   }
> 
>   const result = await users.createOne({
>     first_name: firstName,
>     last_name: lastName,
>     email: email,
>     password: password,
>     role: '80634947-7f0f-497e-a748-fa9ca93dab2a',
>   })
> 
>   res.status(201).json({ message: 'success' })
> 
>   return result
> }
> 
> export default handler
> ```
> 
> and I called it from the front side with a simple function :
> 
> ```js
>     const createUser = async (e, email, password, firstName, lastName) => {
>         e.preventDefault()
>         console.log(email, password, firstName, lastName)
>         const response = await fetch('/api/auth/signup', {
>             method: 'POST',
>             body: JSON.stringify({ email, password, firstName, lastName }),
>             headers: {
>                 "Content-Type": 'application/json'
>             }
>         })
> 
>         const data = await response.json()
> 
>         setErrorMessage(data.message)
> 
>         console.log(errorMessage)
> 
>         return data;
> 
>     }
> ```
> 
> As this way it worked very well ,
> 
> but the login process doesn't work well because of Refresh Tokens , I cant understand it I created a front-end function that handle login like that :
> 
> ```js
>     const loginUser = async (e, email, password, firstName, lastName) => {
>         e.preventDefault()
>         const result = directus.auth.login({ email, password })
>             .then(
>                 () => {
>                     setSuccessMessage("Logged In Successfully")
>                     redirect()
>                 }
>             )
>             .catch((error) => {
>                 console.error(error);
>                 setErrorMessage("Incorrect Data")
>             });
>         return result
>     }
> ```
> 
> I can't understand how to implement refresh token after that
> 
> I created a global state in the context to support the "authenticated" value to the all pages of nextjs :
> 
> ```js
> if (directus.auth.token) {
>   authenticated = true
> }
> 
> setTimeout(function () {
>   const token = directus.auth
>     .refresh()
>     .then(() => {
>       authenticated = true
>     })
>     .catch((error) => {
>       console.log(error)
>       authenticated = false
>     })
> }, 600000)
> ```
> 
> but it gives me error when trying to logout using : directus.auth.logout()
> 
> Cookie “refresh_token” has been rejected because it is already expired.
> 
> Object { message: ""refresh_token" is required in either the JSON payload or Cookie", extensions: {…} }

After using this way , after user signup the directus SDK store a refresh token in the client browser , what can a Cracker do with this 'refresh token ' ?

Also notice that the user i created for registration is limited to only create users , with static token  .

**Reply by alsherif-khalaf** [reply]
Also if the cracker can get a token from this refresh token he will be able to get all users .

**Reply by alsherif-khalaf** [reply]
the refresh token : 

![Screenshot 2022-08-28 173942](https://user-images.githubusercontent.com/73169053/187082424-48b3c6d6-335d-4077-8fab-a17fb9762c3b.png)


How can I prevent Directus SDK from storing refresh token in the client side ?

**Reply by rijkvanzanten** [reply]
@alsherif-khalaf You can't you will have to have some sort of refresh token on the client side in order for the client side to be able to refresh their session. In this particular case, by configuring the cookie with HttpOnly, SameSite=strict, and Secure you can prevent most mitm attacks on that cookie 👍🏻

**Reply by alsherif-khalaf** [reply]
> @alsherif-khalaf You can't you will have to have some sort of refresh token on the client side in order for the client side to be able to refresh their session. In this particular case, by configuring the cookie with HttpOnly, SameSite=strict, and Secure you can prevent most mitm attacks on that cookie 👍🏻

SameSite=strict

How can I do this and the public registration done form another domain , if I subscribed to Directus Cloud with another domain ?

**Reply by rijkvanzanten** [reply]
Ah that wasn't clear to me from your initial question. In that case, you're all set 👍🏻

**Comment by abdonrd** [comment]
After seeing the @rijkvanzanten comment: https://github.com/directus/directus/discussions/11884#discussioncomment-2266781

When you want to have a record open, you should not want the user to customize the role.

So the recommended way is to set permissions to the Public role? Just `email` & `password`.

<img width="857" alt="Screen Shot 2022-04-01 at 19 16 00" src="https://user-images.githubusercontent.com/1007051/161311667-779e998b-639b-43f5-898a-20258acd53d2.png">


And then something like this should works:

```javascript
import axios from 'axios';

axios
  .post('http://localhost:8055/users', {
    email: 'test@test.com',
    password: 'Password',
  })
  .then(response => {
    console.log('User profile', response.data.data);
  })
  .catch(error => {
    console.log('An error occurred:', error.response);
  });
```

**Reply by rijkvanzanten** [reply]
Combined with that, you can use the "Field Presets" tab up above to set a default role for those created users 🙂

**Reply by abdonrd** [reply]
Should be nice to have a feature in the Directus config to allow this kind of public registration.

- [x] Allow public registration

With this feature enabled, when a new user registers, the user will have an `Authenticated` default role (like the `Public` or `Administrator`) assigned automatically.

**Reply by rijkvanzanten** [reply]
Yeah! You're imagining it would also include UI to actually register in the app then yeah? Otherwise you'd effectively end up with the same support as what you've configured above, but as a dedicated endpoint 🙂  (also fwiw, there's another discussion about this: https://github.com/directus/directus/discussions/3083)

**Reply by abdonrd** [reply]
Mmm nop. The UI part would be done in my website/app, not in Directus.

Oh! I had not seen that discussion! 👀
